### ISSUE TITLE
RBAC Filter log sampling

### ISSUE THREAD

**Issue by pateldrashti3** [envoyproxy/envoy#30006]
**If you are reporting *any* crash or *any* potential security issue, *do not*
open an issue in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.**

*Title*: We want to support RBAC filter log sampling and wanted community's opinions

*Description*:
- Envoy's default access log sampling will not work in our case as it will be too many logs logged and miss some important log info for our case, so we want to introduce log sampling just for RBAC Filter. 

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.

**Comment by kyessenov** [comment]
Can you clarify a bit more what you want in the sampling? There's a hint already that could be used to filter RBAC entries: https://www.envoyproxy.io/docs/envoy/latest/configuration/advanced/well_known_dynamic_metadata#shared-dynamic-metadata.

**Comment by yangminzhu** [comment]
Hi, I work with @pateldrashti3 and want to provide a bit more context about this. We're from Uber and planning to use the Envoy RBAC filter. 

We want to log the caller's identity (such as the x509 URI SAN in TLS cert) when a request is allowed or denied by the RBAC filter, there are several requirements:
1) we need sampled logging (e.g. log only 1% of the requests) to control the log size, the log should be avaiable at the info level so we do not have to enable the `debug` log level of the rbac filter all the time;
2) we want to ensure every caller is logged at least once, there could be callers that are making requests occasically and we do not want to miss them due to the sampled logging;
3) we need to put some other internal information in the log output, these information are generated using custom parsing logic we added to the rbac filter internally;

The goal here is to see if there is a more general sampled logging support that can be used by each filter in Envoy, for example, it could be a function like `shouldLog(key string, rate double)` where `key` is used to decide if the request has been logged before and the `rate` decides the frequency of the sampled logging. This would allow us to simplify and minimize the custom changes we need in the rbac filter.

The `access_log_hint` seems still very limited and won't support our use cases:
1) it is only set by the `RBAC::LOG` action for matched access, but we need it for other actions and both matched/non-matched cases;
1) More importantly, it does not seem to have a way to ensure the log at least once semantics?
2) Last, I'm not sure if there is a good way to allow passing custom string from the rbac filter to the access logger in this case;

**Comment by kyessenov** [comment]
Thanks for explaining.

Re 2: There was a discussion in the past on a stateful [log filter](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/accesslog/v3/accesslog.proto#envoy-v3-api-msg-config-accesslog-v3-accesslogfilter), similar to the rate limit 
 configuration. I imagine you want to log once per source URI SAN every minute, for example. I think a matcher could work here, that produces a string "key" per request, e.g. see https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/rate_limit_quota_filter.html#example-1, plus a token bucket algorithm. All the primitives are there IMHO.

Re 1-3: There was work to add a dedicated audit logger to RBAC in both gRPC and Envoy, please see https://github.com/envoyproxy/envoy/pull/26453. 

I can't tell what info you are parsing, but filter state or metadata is usually necessary to pass the logging info the logger from a filter.

**Comment by kyessenov** [comment]
cc @markdroth since there was some recent activity on audit logging xDS.

**Comment by markdroth** [comment]
Yeah, I think the audit logger API that we added in #26001 is the right way to solve this.  It was designed explicitly for audit logging in RBAC, and we're in the process of implementing this in gRPC.  So you could address this issue by implementing that API in Envoy, and then writing a plugin for that API that does the appropriate sampling and logs the info you want.

**Comment by yangminzhu** [comment]
@kyessenov  thank you very much for the suggestion, that's super helpful! 

>  I imagine you want to log once per source URI SAN every minute, for example. I think a matcher could work here, that produces a string "key" per request, e.g. see https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/rate_limit_quota_filter.html#example-1, plus a token bucket algorithm. All the primitives are there IMHO.

yes, we want to do the sampled log based on the source URI SAN. The unified matcher can be used to produce the key with the source URI SAN, but we will still need some code changes to implement the `log at least once per-key` sampling  which is not supported in the `accesslog` today, is that right? Just want to check to make sure I didn't miss anything here.

@markdroth  the audit logger API seems also promising for our use case. but I guess we will need to extend the audit logger (such as `StdoutAuditLog`) to support the sampled logging based on some key (such as the source URI SAN)? Also do you know is anyone working on implementing the API in the Envoy side? It looks like the API is added half a year ago but not implemented yet, so wondering what's the plan here.

In summary, it seems we will need code changes either way:
- If we use the access log, we need to implement the `log at least once per-key` sampling logic in the `accesslog`. Note as we're not using HCM, we need to configure access log on the listener (I assume we couldn't configure it on the `tcp_proxy` because when a connection is rejected by RBAC, the `tcp_proxy` filter will not be invoked at all).

- If we use the audit logger API in RBAC, we can configure everything inside the RBAC filter. The audit log API still needs to be implemented plus the extra sampling logic.
